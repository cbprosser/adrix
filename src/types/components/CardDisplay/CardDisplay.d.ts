import { APICard, APICardFace, APILayout } from '../../';

export interface CardDisplayProps {
  card?: APICard;
}

export interface CardFaceDisplayProps {
  face: APICardFace;
}

export interface CardImageDisplayProps {
  faces: APICardFace[];
  layout?: APILayout;
}

export interface CardImageDisplayState {
  face: 0 | 1;
}
