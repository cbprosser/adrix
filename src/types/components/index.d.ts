export * from './AdvancedSearchForm';
export * from './MTGStaticTypes';
export * from './MultipleCardDisplay';
export * from './SimpleCardSearchBar';
export * from './Symbol';
