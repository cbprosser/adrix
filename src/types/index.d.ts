export * from './axios';
export * from './components';
export * from './models';
export * from './redux';
export * from './sitemap';
