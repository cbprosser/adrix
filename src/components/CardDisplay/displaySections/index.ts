export * from './CardFace';
export * from './CardImage';
