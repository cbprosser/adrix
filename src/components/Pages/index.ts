export * from './AdvancedSearch';
export * from './Home';
export * from './Results';
