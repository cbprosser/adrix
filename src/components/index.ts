export * from './AppHeader';
export * from './CardDisplay';
export * from './Pages';
export * from './SimpleCardSearchBar';
