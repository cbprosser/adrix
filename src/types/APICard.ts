import { layouts, rarities, relatedComponents } from '../util';

/* eslint-disable camelcase */
export type APICard = APICore & APIGameplay & APIPrint;

export type APICore = {
  arena_id?: number;
  id: APIUUID;
  lang: string; // TODO: literalise
  mtgo_id?: number;
  mtgo_foil_id?: number;
  multiverse_ids?: number[];
  tcgplayer_id?: number;
  cardmarket_id?: number;
  object: 'card';
  oracle_id: APIUUID;
  prints_search_uri: APIURI;
  rulings_uri: APIURI;
  scryfall_uri: APIURI;
  uri: APIURI;
};

export type APIGameplay = {
  all_parts?: APIAllParts[];
  card_faces?: APICardFace[];
  cmc: number;
  color_identity: APIColors[];
  color_indicator?: APIColors[];
  colors?: APIColors[];
  edhrec_rank?: number;
  foil: boolean;
  hand_modifier?: string;
  keywords: string[];
  layout: APILayout;
  legalities: APILegalities;
  life_modifier?: string;
  loyalty?: string;
  mana_cost?: string;
  name: string;
  nonfoil: boolean;
  oracle_text?: string;
  oversized: boolean;
  power?: string;
  produced_mana?: APIColors[];
  reserved: boolean;
  toughness?: string;
  type_line: string;
};

export type APIPrint = {
  artist?: string;
  booster: boolean;
  border_color: string;
  card_back_id: APIUUID;
  collector_number: string;
  content_warning?: boolean;
  digital: boolean;
  flavor_name?: string;
  flavor_text?: string;
  frame_effects?: string[]; // TODO: literalise
  frame: string;
  full_art: boolean;
  games: string[]; // TODO: literalise
  highres_image: boolean;
  illustration_id?: APIUUID;
  image_status: string; // TODO: literalise
  image_uris?: APIImageURIS;
  prices: APIPrices;
  printed_name?: string;
  printed_text?: string;
  printed_type_line?: string;
  promo: boolean;
  promo_types?: string[];
  purchase_uris: APIPurchaseURIS;
  rarity: APIRarity;
  related_uris: APIRelatedURIS;
  released_at: APIDateString;
  reprint: boolean;
  scryfall_set_uri: APIURI;
  set_name: string;
  set_search_uri: APIURI;
  set_type: string;
  set_uri: APIURI;
  set: string;
  story_spotlight: boolean;
  textless: boolean;
  variation: boolean;
  variation_of?: APIUUID;
  watermark?: string;
  preview?: APIPreview;
};

export type APICardFace = {
  artist?: string;
  color_indicator?: string;
  colors?: APIColors[];
  flavor_text?: APIColors[];
  illustration_id?: APIUUID;
  image_uris?: APIImageURIS;
  loyalty?: string;
  mana_cost: string;
  name: string;
  object: string;
  oracle_text?: string;
  power?: string;
  printed_name?: string;
  printed_text?: string;
  printed_type_line?: string;
  toughness?: string;
  type_line: string;
  watermark?: string;
};

export type APIAllParts = {
  id: APIUUID;
  object: 'related_card';
  component: APIRelatedComponents;
  name: string;
  type_line: string;
  uri: APIURI;
};

export type APIRelatedURIS = {
  gatherer: APIURI;
  [key: string]: APIURI;
};

export type APIPurchaseURIS = {
  tcgplayer: APIURI;
  cardmarket: APIURI;
  cardhoarder: APIURI;
  [key: string]: APIURI;
};

export type APIPrices = {
  usd?: string;
  usd_foil?: string;
  eur?: string;
  eur_foil?: string;
  tix?: string;
};

export type APIImageURIS = {
  small?: APIURI;
  normal?: APIURI;
  large?: APIURI;
  png?: APIURI;
  art_crop?: APIURI;
  border_crop?: APIURI;
};

export type APIPreview = {
  previewed_at?: APIDateString;
  source_uri?: APIURI;
  source?: string;
};

export type APIColors = 'W' | 'U' | 'B' | 'R' | 'G';

export type APILegalities = {
  [key: string]: 'legal' | 'not legal' | 'restricted' | 'banned';
};

export type APIURI = string;

export type APIUUID = string;

export type APIDateString = string;

export type APILayout = typeof layouts[number];

export type APIRarity = typeof rarities[number];

export type APIRelatedComponents = typeof relatedComponents[number];
