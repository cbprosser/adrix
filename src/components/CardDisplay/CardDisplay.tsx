import { Box, Divider, Paper, styled, Typography } from '@mui/material';
import { Fragment } from 'react';
import { APICard } from '../../types';
import { convertSymbols } from '../../util';

interface CardDisplayProps {
  card?: APICard;
}

const SpacedDivider = styled(Divider)(({ theme }) => ({
  marginTop: theme.spacing(0.5),
  marginBottom: theme.spacing(0.5),
  marginLeft: theme.spacing(-1),
  marginRight: theme.spacing(-1),
}));

export const CardDisplay = ({ card }: CardDisplayProps) => {
  const {
    name,
    image_uris,
    mana_cost,
    type_line,
    oracle_text,
    power,
    toughness,
    loyalty,
    artist,
  } = card || {};

  const { normal } = image_uris ?? {};

  return card !== undefined ? (
    <Paper elevation={0} sx={{ display: 'flex' }}>
      <Box component="img" src={normal} sx={{ maxHeight: 400 }} />
      <Paper sx={(theme) => ({ padding: theme.spacing(1), width: '33.33%' })}>
        <Typography>
          {name}
          {mana_cost ? [' - ', convertSymbols(mana_cost)] : undefined}
        </Typography>
        <SpacedDivider />
        <Typography>{type_line}</Typography>
        <SpacedDivider />
        {oracle_text?.split('\n').map((line, i) => (
          <Typography key={i}>{convertSymbols(line)}</Typography>
        ))}
        <SpacedDivider />
        {(loyalty !== undefined || power !== undefined) && (
          <Fragment key={loyalty ? 'loyalty' : 'pt'}>
            <Typography>{loyalty ?? `${power}/${toughness}`}</Typography>
            <SpacedDivider />
          </Fragment>
        )}
        <Typography>{artist}</Typography>
      </Paper>
      <Paper elevation={0} sx={{ flexGrow: 1 }}>
        coming soon
      </Paper>
    </Paper>
  ) : (
    <></>
  );
};
