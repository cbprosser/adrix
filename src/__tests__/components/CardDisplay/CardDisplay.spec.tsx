import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CardDisplay } from '../../../components/CardDisplay';
import { APICard } from '../../../types';
import {
  mockCard,
  mockFlipCard,
  mockPlaneswalkerCard,
  mockTransformCard,
} from '../../testUtils';

const waitFor = (timeout = 1000) =>
  new Promise((res) => setTimeout(() => res(undefined), timeout));

const user = userEvent.setup();

const renderComponent = (card?: APICard) => render(<CardDisplay card={card} />);

describe('CardDisplay Suite', () => {
  it('Should match the snapshot when no card is present', () => {
    const view = renderComponent();

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('Should match the snapshot when a standard card is present', () => {
    const view = renderComponent(mockCard);

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('Should match the snapshot when a planeswalker card is present', () => {
    const view = renderComponent(mockPlaneswalkerCard);

    expect(view.asFragment()).toMatchSnapshot();
  });

  describe('Multi faced card tests', () => {
    const renderTransform = () => renderComponent(mockTransformCard);
    it('Should match the snapshot when a multi faced card is present', () => {
      const view = renderTransform();

      expect(view.asFragment()).toMatchSnapshot();
    });

    it('Should have a transform button', async () => {
      renderTransform();

      const buttons = await screen.findAllByRole('button');

      expect(buttons.length).toBeGreaterThan(0);

      const transformButton = buttons[0];

      const svg = await within(transformButton).findByTestId('LoopIcon');

      expect(svg).toBeInTheDocument();
    });

    it('Should flip the card when the transform button is clicked', async () => {
      const { asFragment } = renderTransform();

      const buttons = await screen.findAllByRole('button');

      expect(buttons.length).toBeGreaterThan(0);

      const transformButton = buttons[0];

      const initialFragment = asFragment();
      await user.click(transformButton);
      await waitFor();
      expect(initialFragment).toMatchDiffSnapshot(asFragment());
      await user.click(transformButton);
      await waitFor();
      expect(initialFragment).toMatchDiffSnapshot(asFragment());
    });
  });

  describe('Multi sided card test', () => {
    const renderFlip = () => render(<CardDisplay card={mockFlipCard} />);
    it('Should match the snapshot when a multi sided card is present', () => {
      const { asFragment } = renderFlip();

      expect(asFragment()).toMatchSnapshot();
    });

    it('Should have a transform button', async () => {
      renderFlip();

      const buttons = await screen.findAllByRole('button');

      expect(buttons.length).toBeGreaterThan(0);

      const transformButton = buttons[0];

      const svg = await within(transformButton).findByTestId('LoopIcon');

      expect(svg).toBeInTheDocument();
    });

    it('Should flip the card when the transform button is clicked', async () => {
      const { asFragment } = renderFlip();

      const buttons = await screen.findAllByRole('button');

      expect(buttons.length).toBeGreaterThan(0);

      const transformButton = buttons[0];

      const initialFragment = asFragment();
      await user.click(transformButton);
      await waitFor();
      expect(initialFragment).toMatchDiffSnapshot(asFragment());
      await user.click(transformButton);
      await waitFor();
      expect(initialFragment).toMatchDiffSnapshot(asFragment());
    });
  });
});
