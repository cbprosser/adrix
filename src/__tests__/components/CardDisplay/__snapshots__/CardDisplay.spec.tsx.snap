// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDisplay Suite Should match the snapshot when a planeswalker card is present 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-0-MuiPaper-root"
  >
    <img
      class="MuiBox-root css-1"
      src="https://c1.scryfall.com/file/scryfall-cards/normal/front/1/7/178bf300-38cf-4ab2-9e66-6fcaa112b649.jpg?1627429348"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-2-MuiPaper-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        Kasmina, Enigma Sage - {1}{G}{U}
      </p>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-4-MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        Legendary Planeswalker — Kasmina
      </p>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-4-MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        Each other planeswalker you control has the loyalty abilities of Kasmina, Enigma Sage.
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        +2: Scry 1.
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        −X: Create a 0/0 green and blue Fractal creature token. Put X +1/+1 counters on it.
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        −8: Search your library for an instant or sorcery card that shares a color with this planeswalker, exile that card, then shuffle. You may cast that card without paying its mana cost.
      </p>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-4-MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        2
      </p>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-4-MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        Tyler Jacobson
      </p>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-15-MuiPaper-root"
    >
      coming soon
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardDisplay Suite Should match the snapshot when a standard card is present 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-0-MuiPaper-root"
  >
    <img
      class="MuiBox-root css-1"
      src="https://c1.scryfall.com/file/scryfall-cards/normal/front/d/4/d495e084-92ec-4675-b6aa-8ade2775b796.jpg?1625190786"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-2-MuiPaper-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        Adrix and Nev, Twincasters - {2}{G}{U}
      </p>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-4-MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        Legendary Creature — Merfolk Wizard
      </p>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-4-MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        Ward {2} (Whenever this creature becomes the target of a spell or ability an opponent controls, counter it unless that player pays {2}.)
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        If one or more tokens would be created under your control, twice that many of those tokens are created instead.
      </p>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-4-MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        2/2
      </p>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-4-MuiDivider-root"
      />
      <p
        class="MuiTypography-root MuiTypography-body1 css-3-MuiTypography-root"
      >
        Andrew Mar
      </p>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-13-MuiPaper-root"
    >
      coming soon
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardDisplay Suite Should match the snapshot when no card is present 1`] = `<DocumentFragment />`;
